framework/src/kernels/KernelScalarBase.C

```diff
@@ -0,0 +1,258 @@
+          _local_ke(_h, _j) += _JxW[_qp] * _coord[_qp] * computeScalarQpOffDiagJacobian(jvar_num);
+    }
+  }
+  else if (jvar.fieldType() == Moose::VarFieldType::VAR_FIELD_ARRAY)
+    mooseError("Array variable cannot be coupled into Kernel Scalar currently");
+  else
+    mooseError("Vector variable cannot be coupled into Kernel Scalar currently");
+
+  for (const auto & matrix_tag : _matrix_tags)
+    _assembly.cacheJacobianBlock(_local_ke,
+                                 _kappa_var_ptr->dofIndices(),
+                                 jvar.dofIndices(),
+                                 _kappa_var_ptr->scalingFactor(),
+                                 matrix_tag);
+}
+
+void
+KernelScalarBase::computeOffDiagJacobianScalarLocal(const unsigned int svar_num)
+{
+
+  // Get dofs and order of this scalar; at least one will be _kappa_var
+  const auto & svar = _sys.getScalarVariable(_tid, svar_num);
+  const unsigned int s_order = svar.order();
+  _local_ke.resize(_test.size(), s_order);
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    initScalarQpJacobian(svar_num);
+    for (_i = 0; _i < _test.size(); _i++)
+      for (_l = 0; _l < s_order; _l++)
+        _local_ke(_i, _l) += _JxW[_qp] * _coord[_qp] * computeQpOffDiagJacobianScalar(svar_num);
+  }
+
+  for (const auto & matrix_tag : _matrix_tags)
+    _assembly.cacheJacobianBlock(
+        _local_ke, _var.dofIndices(), svar.dofIndices(), _var.scalingFactor(), matrix_tag);
+}
+
+void
+KernelScalarBase::computeOffDiagJacobianScalar(const unsigned int svar_num)
+{
+  precalculateOffDiagJacobianScalar(svar_num);
+  if (_use_scalar)
+  {
+    if (svar_num == variable().number()) // column for this kernel's variable
+    {
+      // this kernel's variable is not a scalar
+      return;
+    }
+    else if (svar_num == _kappa_var) // column for this kernel's scalar variable
+    {
+      // Perform assembly using method in Kernel; works for simple cases but not general
+      // Kernel::computeOffDiagJacobianScalar(svar_num); // d-_var-residual / d-_kappa
+      // Perform assembly using local_ke like d-_kappa_var-residual / d-_var
+      if (_compute_field_residuals)
+        computeOffDiagJacobianScalarLocal(svar_num); // d-_var-residual / d-_kappa
+      if (_compute_scalar_residuals)
+        computeScalarJacobian(); // d-_kappa-residual / d-_kappa
+    }
+    else // some other column for scalar variable
+    {
+      // Perform assembly using method in Kernel; works for simple cases but not general
+      // Kernel::computeOffDiagJacobianScalar(svar_num); // d-_var-residual / d-jvar
+      // Perform assembly using local_ke like d-_kappa_var-residual / d-_var
+      if (_compute_field_residuals)
+        computeOffDiagJacobianScalarLocal(svar_num); // d-_var-residual / d-svar
+      if (_compute_scalar_residuals)
+        computeScalarOffDiagJacobianScalar(svar_num); // d-_kappa-residual / d-svar
+    }
+  }
+  else if (_compute_field_residuals)
```

Comment:
Should this be an `if` instead of an `else if`?

Reply:
In short, I feel it's a `else if`.

If I learned C++ correctly, then the switch-case branches will go through all cases unless a break is given, while if-elseif-else only goes past the first 'if' if that conditional is false.
The logic of the statement above is, if there is a scalar variable (`_use_scalar == true`), then we do all the stuff above. If there is not a scalar, then we just want to do the `Kernel::computeOffDiagJacobianScalar` normal method like it would in a typical `Kernel`. Yet, the new functionality added by the parameters is that the `_compute_field_residuals` is controlling if any assembly happens.
If that parameter as well as `_use_scalar` are false, then this `computeOffDiagJacobianScalar` should behave like `{}` and do nothing.